pyomo/solvers/plugins/solvers/scip_direct.py

```diff
@@ -0,0 +1,838 @@
+        vtype = self._scip_vtype_from_var(var)
+        lb, ub = self._scip_lb_ub_from_var(var)
+
+        scip_var = self._solver_model.addVar(lb=lb, ub=ub, vtype=vtype, name=varname)
+
+        self._pyomo_var_to_solver_var_expr_map[var] = scip_var
+        self._pyomo_var_to_solver_var_map[var] = scip_var.name
+        self._solver_var_to_pyomo_var_map[varname] = var
+        self._referenced_variables[var] = 0
+
+    def close(self):
+        """Frees SCIP resources used by this solver instance."""
+
+        if self._solver_model is not None:
+            self._solver_model.freeProb()
+            self._solver_model = None
+
+    def __exit__(self, t, v, traceback):
+        super().__exit__(t, v, traceback)
+        self.close()
+
+    def _set_instance(self, model, kwds={}):
+        DirectOrPersistentSolver._set_instance(self, model, kwds)
+        try:
+            self._solver_model = self._scip.Model()
+        except Exception:
+            e = sys.exc_info()[1]
+            msg = (
+                "Unable to create SCIP model. "
+                "Have you installed PySCIPOpt correctly?\n\n\t"
+                + "Error message: {0}".format(e)
+            )
+            raise Exception(msg)
+
+        self._add_block(model)
+
+        for var, n_ref in self._referenced_variables.items():
+            if n_ref != 0:
+                if var.fixed:
+                    if not self._output_fixed_variable_bounds:
+                        raise ValueError(
+                            "Encountered a fixed variable (%s) inside "
+                            "an active objective or constraint "
+                            "expression on model %s, which is usually "
+                            "indicative of a preprocessing error. Use "
+                            "the IO-option 'output_fixed_variable_bounds=True' "
+                            "to suppress this error and fix the variable "
+                            "by overwriting its bounds in the SCIP instance."
+                            % (var.name, self._pyomo_model.name)
+                        )
+
+    def _add_block(self, block):
+        DirectOrPersistentSolver._add_block(self, block)
+
+    def _add_constraint(self, con):
+        if not con.active:
+            return None
+
+        if is_fixed(con.body) and self._skip_trivial_constraints:
+            return None
+
+        conname = self._symbol_map.getSymbol(con, self._labeler)
+
+        if con._linear_canonical_form:
+            scip_expr, referenced_vars = self._get_expr_from_pyomo_repn(
+                con.canonical_form(), self._max_constraint_degree
+            )
+        else:
+            scip_expr, referenced_vars = self._get_expr_from_pyomo_expr(
+                con.body, self._max_constraint_degree
+            )
+
+        if con.has_lb():
+            if not is_fixed(con.lower):
+                raise ValueError(
+                    "Lower bound of constraint {0} is not constant.".format(con)
```

Comment:
A general note: This file alone switches between f-strings, `str.format`, and conversion specifiers. It's kind of weird to see all used in a single file.

Reply:
This is my bad on copying code from the other persistent interfaces, and probably someone else's bad for copying code in turn (I assume that's why there's three standards). 
I am guilty of all the f-strings, and prefer that format, so have changed everything in scip_x.py to match.

Reply:
Copy-pasta catches us all.  🤷‍♀️